Implement item-spheres + Last WotH

damdayou opened the pull request:
This adds a function to compute item-spheres the same way that players would track item-spheres and print them in the spoiler log. These item-spheres are coarser than the ones currently listed in the playthrough but made to be easy to track on player side.

We then use these item-spheres to correct matthewkirby/DGBarca's Last WotH Hint system so that it takes from the furthest location according to these item-spheres rather than the current WotH location list whose order could provide the wrong Last Woth.

Special thanks goes to ArthurOudini for his intensive testing and precious feedback.

TreZc0 commented:
Please provide us with a more extensive explanation how both new functionalities work.

Especially the sphere system has been questioned in the past, so building something on top does create a few questions.

it would help a lot to have an indepth explanation of everything this does, and the benefit you expect from this being merged.

damdayou commented:
Thank you for all your answers and sorry for the late reply.

Let me first explain the "why" of this PR. This project originated with Arthur, others and I studying various spoiler logs to understand how required location list and playthrough spheres are computed. After facing many cases that didn't match our intuition, we ended up looking into the randomizer source code. We learned that:
1. playthrough sphere "increments" can be triggered by picking litterally anything (keys, consumables, etc.), including some technical _subrules_ that are invisible to the player,
2. the order of the required location list is not guaranteed to be consistant with the playthrough (because they are computed by two independent functions, namely `update_required_items()` and `create_playthrough()`).

Despite being totally fine in terms of seed generation, we were not happy with this behavior, having in mind Last WotH hint implementation. Issue 1 makes playthrough spheres almost impossible to live-track for a player (making Last WotH barely more useful than a regular WotH). Issue 2 makes it impossible to implement Last WotH by simply reading the required location list backwards (#1307 used to do that and provided a wrong Last WotH in some scenarios). **The main goal of this PR is then to add a new sphere system much simpler to track than playthrough spheres, that we decided to call _item spheres_.**

These new item spheres are computed by the function `compute_coarse_spheres()`, which is run after `create_playthrough()` and before `update_required_items()`. This function creates an independent item sphere list. **It leaves the current playthrough spheres and required location list untouched, because we did not want to interfere with seed generation.** We also modified the `to_json()` method of the `Distribution` class so that the item spheres appear in a dedicated section in the spoiler log.

I can elaborate on the `compute_coarse_sphere()` function if needed, but I feel like you already got what it does. This is basically the same algorithm as in `update_required_items()`, except that only _noteworthy_ items trigger sphere jumps (and that makes a _huge_ difference). This means that a player who would want to track those new item spheres only needs to track these noteworthy items and can ignore any other item. That said, our main task was to define a set of noteworthy items that is both simple to track while playing a seed and wide enough to produce interesting spheres in terms of seed logic. The rule to determine if an item is notheworthy or not can be found in detail in the function `is_noteworthy()`, but basically it is simply: items and songs that would appear in a typical item tracker. 

After having tested it out ourselves on many seeds, it seems to us that **an average player can track these new item spheres while playing** (I could do it despite being quite a low-level player). Of course I don't claim it requires no effort nor that you will be right 100% of the time, but it requires more or less the same knowledge as for regular WotH and as such rewards players for their understanding of a seed logic.

Concerning the Last WotH hint, it is one possible application of the item spheres implementation (which, once you track item spheres, becomes much more subtle and useful than just giving information on the "last" item) but I am convinced it is far from being the only one. If I remember well, when we first introduced this PR on the OoTR Discord, some users quickly suggested other uses, such as a hint system that would try to distribute WotH more equally between early-game, mid-game and late-game sphere. I would be really happy to see people thinking and implementing new hint distributions (or more?) on top of these item spheres.

This message is already long but I'd like to answer some of your remarks.

> One issue i do see is that is fully standard focused, while not being able to limit it to Standard.

The current algorithm is indeed standard focused because these are the settings we had in mind when writing it. But if it appears that the set of _noteworthy_ items we chose does not perform well for some other settings, I guess it is possible to adjust this set either by making this set setting dependant, or by slightly modifying it so that it becomes general enough (I don't have enough experience in other settings than standard to do it myself, though).

> All rulesets that use glitchless logic, but allow glitches, would potentially get wrong information out of this hint type.

If feel like this already applies to classical WotH hints, and that being able to play "out-of-logic" while trying to keep track of the "seed logic" is already common (even in standard with chus, skips, etc) and interesting in rando. But maybe I misunderstood what you meant, let me know.

> I think 'last woth' hint type is a bad idea. It requires you to know the precise details of how it was calculated, and then parse the total logic of everything in the seed, to use it effectively. I don't believe this can realistically be done without a logic tracker. But, if you can use it to its full potential, you get such a huge amount of information on where items can even be. I think it would basically force all players to use a logic tracking tool.

I'd like to insist on the fact that the item spheres we implemented are way much simpler than the ones in the playthrough. The way they are calculated can be summed up as follows:
- sphere -1 is what you have immediately when starting the file,
- sphere 0 is what you can collect without any other item on your tracker,
- sphere N+1 is what can be collected using only sphere ≤N items on your tracker (including one item in sphere N exactly).

Item spheres tracking does not require a logic tracker. You can track item spheres on a piece of paper or on a tracker with some empty drag'n drop section (like the one you can see in [Arthur's video](https://www.youtube.com/watch?v=XUyyqJ1Nx6Q)). And regarding the Last WotH hint, it only helps ruling out some areas in late game that are locked by items in the sphere containing your last WotH item. From our experience, it has variable usefulness (from "as useful as any other WotH" to "ruling out a couple full areas"), but I fail to see how it could be as powerful as giving the precise location of a specific item without having cleared an already substantial part of the seed.

> I think Last WOTH can be done/explained without the use of spheres, as in without the current playthrough spheres, the "area+age-spheres" in an experimental branch, or a new item-spheres. Name-wise, I'd look for something more flavorful like "hero's final quest" or "hero's penultimate prize".

One advantage I see of building Last WotH item on top of spheres is that Last WotH becomes more than just giving you that last item you need to collect. If you track your spheres correctly, you know that _any item in the same sphere as what you think is your last Woth item_ has the same property as your last WotH item. This potentially rules out more locations from your late game research than just the last WotH item alone.

TestRunnerSRL commented:
Hmm. I've talked about this on the Discord a bit before, but I'll just put some of my thoughts here as well. I feel like you are doing a bit of misrepresenting how the current sphere algorithm works, which is part of the reason this is confusing to me. The current alg doesn't "increment" sphere on arbitrary items at all. The algorithm scans the world for all reachable new locations, and add that to a sphere. The sphere is "incremented" when the search runs out of locations. It then "collects" the scanned items and repeats with a new sphere. This process repeats until all items (or end goal or wtv). So the only real rule that's being enforced is that:

1) Having all items from sphere 0-N will allow obtaining all items in N+1
2) Having only items from sphere 0-N will not allow obtaining any items from N+2

There's going to be inherit issues trying to give a proper ordering to the item list because many items can be gotten in very different orders. Even outside of a single sphere, you can get items from sphere 12, while still be missing items in sphere 1. Because of how the fill alg works, it's generally not very important to know what "sphere" number items fall under, because that is just not how generation works. No matter what sets of locations from sets of spheres, the remaining unchecked accessible locations generally have the same chances of having the same items (barring location specific restrictions).

Ok, with that out of the way, this "item spheres" algorithm seems to be breaking up the spheres in somewhat strange ways. Instead of splitting spheres based on when the search runs out, it seems to kinda merge spheres together that don't have significant progress (which is sorta reasonable), and then split in the middle of the sphere when a major item is found (which is not really reasonable). The order the items show up in the search is effectively random, and therefore you'll run into situations for items might be in an earlier or later sphere completely dependent on the order locations get scanned. This can also lead to what looks like an ordering of the significant items, when once again they may be able to be gotten out of order still.

Also as noted, the set for "notable items" is extremely settings dependent, which is something in general we avoided having. New settings need to be functional is all/most settings, or we'd have to deal with some integration hell somewhere down the road.

In terms of player tracking, this is not really information that the player normally needs to know at all. The playthrough algorithm isn't used for any hint generation for pretty strong reasons. There's two functions you pointed out because the playthrough is there for logging/spoilers, while the required_items is actually useful hint information. Beyond that as mentioned, the playthrough requires the player to understand exactly how the playthrough is generated, and all the order biases that come from the search algorithm. And similarly for determining "other locations in the final sphere" is information that is approximate at best. As said, you don't know which items may be in a slightly later or earlier item sphere, and it's still dependent on the order the search algorithm runs.

As for how useful "last woth" is, I really can't say. Some people seem to be kinda wary about the usefulness of such a type of hint, but I don't really know atm. I definitely agree that a "last woth" can be implemented without this "item sphere" algorithm, but I'm unsure of its value. Also as mentioned, there is some ordering problems where it can be ambiguous which item is the last woth. Frequently many items (including woth) can be obtained interchangeably, which is why it really had no ordering previously. It was only really in a list so that we could shuffle it.

damdayou commented:
>Hmm. I've talked about this on the Discord a bit before, but I'll just put some of my thoughts here as well. I feel like you are doing a bit of misrepresenting how the current sphere algorithm works, which is part of the reason this is confusing to me. The current alg doesn't "increment" sphere on arbitrary items at all. The algorithm scans the world for all reachable new locations, and add that to a sphere. The sphere is "incremented" when the search runs out of locations. It then "collects" the scanned items and repeats with a new sphere. This process repeats until all items (or end goal or wtv).

My bad for this. I realize that I shouldn't have used the word _increment_ which is misleading since it suggests some counter is incremented while the loop is running. Just to be sure we're on the same page, the algorithm in `compute_coarse_sphere()` follows the same algorithm you just described above, up to a small variation. Namely:
 - It scans the world for all reachable new locations, and adds that to a sphere,
 - Once the search runs out of locations:
    - if at least one not-noteworthy item has been scanned, then it collects all the scanned not-noteworthy items and repeats, except that it keeps adding the reachable locations in the same sphere (no sphere "increment"),
   - else it collects all the noteworthy items and repeats with a new sphere.
 - The process repeats until there are no more new reachable locations.

As you pointed out, I guess that in some cases, some playthrough spheres could indeed be entangled among item spheres, but I don't feel like this is an issued since those two sphere lists have distinct uses and are not related.

> There's going to be inherit issues trying to give a proper ordering to the item list because many items can be gotten in very different orders. Even outside of a single sphere, you can get items from sphere 12, while still be missing items in sphere 1.

I fully agree with you on this. The point I failed to make clear is that this PR does not try to order items, but only to order spheres. When an item is hinted as Last WotH, it must no be interpreted as the **last** (hintable) required item that will be collected, but only as being part of the last sphere containing (hintable) required items. Within this "last significant sphere" (which is not the last sphere in general), items can of course be collected in arbitrary order by the player. This is a direct consequence of what we all seem to all agree to be the definition of a "generic" _sphere_.

The remaining question is what precise definition of sphere to take in this PR. Playthrough spheres use the most basic definition (every item/event/subrule is treated the same way) which is perfectly fine for sphere generation but is nearly impossible to track easily for a player. As mentionned before, we strongly believe that the definition we decided to use independently for item spheres (i.e., "ignoring" some items considered as minor) gives something that is live-trackable without too much effort, which is our primary objective, but unfortunately I understand your concerns about adaptability to other settings than standard. However I feel confident about finding something working out well without modifying a lot the current item set. We started studying this question deeper and we will try to come again with a satisfying approach.

cjohnson57 commented:
Sorry, I'm not sure if we want this. The idea is cool and I appreciate the effort (plus finding any way to innovate on the rando algorithms is neat to me), but the interest seems low and not worth the changes. That being said I could be wrong, if people are interested in it feel free to reopen.
